File: frappe/core/page/messages/messages.py
# Copyright (c) 2013, Web Notes Technologies Pvt. Ltd. and Contributors
# MIT License. See license.txt

from __future__ import unicode_literals
import frappe
from frappe.core.doctype.notification_count.notification_count import delete_notification_count_for
from frappe.core.doctype.user.user import STANDARD_USERS

@frappe.whitelist()
def get_list(arg=None):
	"""get list of messages"""
	frappe.form_dict['limit_start'] = int(frappe.form_dict['limit_start'])
	frappe.form_dict['limit_page_length'] = int(frappe.form_dict['limit_page_length'])
	frappe.form_dict['user'] = frappe.session['user']

	# set all messages as read
	frappe.db.begin()
	frappe.db.sql("""UPDATE `tabComment`
	set docstatus = 1 where comment_doctype in ('My Company', 'Message')
	and comment_docname = %s
	""", frappe.user.name)
	
	delete_notification_count_for("Messages")
	
	frappe.db.commit()

	if frappe.form_dict['contact'] == frappe.session['user']:
		# return messages
		return frappe.db.sql("""select * from `tabComment` 
		where (owner=%(contact)s 
			or comment_docname=%(user)s 
			or (owner=comment_docname and ifnull(parenttype, "")!="Assignment"))
		and comment_doctype ='Message'
		order by creation desc
		limit %(limit_start)s, %(limit_page_length)s""", frappe.local.form_dict, as_dict=1)		
	else:
		return frappe.db.sql("""select * from `tabComment` 
		where (owner=%(contact)s and comment_docname=%(user)s)
		or (owner=%(user)s and comment_docname=%(contact)s)
		or (owner=%(contact)s and comment_docname=%(contact)s)
		and comment_doctype ='Message'
		order by creation desc
		limit %(limit_start)s, %(limit_page_length)s""", frappe.local.form_dict, as_dict=1)
		

@frappe.whitelist()
def get_active_users():
	return frappe.db.sql("""select name,
		(select count(*) from tabSessions where user=tabUser.name
			and timediff(now(), lastupdate) < time("01:00:00")) as has_session
	 	from tabUser 
		where ifnull(enabled,0)=1 and
		ifnull(user_type, '')!='Website User' and 
		name not in ({})
		order by first_name""".format(", ".join(["%s"]*len(STANDARD_USERS))), STANDARD_USERS, as_dict=1)

@frappe.whitelist()
def post(arg=None):
	import frappe
	"""post message"""
	if not arg:
		arg = {}
		arg.update(frappe.local.form_dict)
	
	if isinstance(arg, basestring):
		import json
		arg = json.loads(arg)

	d = frappe.new_doc('Comment')
	d.parenttype = arg.get("parenttype")
	d.comment = arg['txt']
	d.comment_docname = arg['contact']
	d.comment_doctype = 'Message'
	d.insert(ignore_permissions=True)
	
	delete_notification_count_for("Messages")

	import frappe.utils
	if frappe.utils.cint(arg.get('notify')):
		notify(arg)
	
@frappe.whitelist()
def delete(arg=None):
	frappe.db.sql("""delete from `tabComment` where name=%s""", 
		frappe.form_dict['name']);

def notify(arg=None):
	from frappe.utils import cstr, get_fullname, get_url
	
	try:
		frappe.sendmail(\
			recipients=[frappe.db.get_value("User", arg["contact"], "email") or arg["contact"]],
			sender= frappe.db.get_value("User", frappe.session.user, "email"),
			subject="New Message from " + get_fullname(frappe.user.name),
			message=frappe.get_template("templates/emails/new_message.html").render({
				"from": get_fullname(frappe.user.name),
				"message": arg['txt'],
				"link": get_url()
			})
		)
	except frappe.OutgoingEmailError, e:
		pass

File: frappe/core/page/messages/messages.js
// Copyright (c) 2013, Web Notes Technologies Pvt. Ltd. and Contributors
// MIT License. See license.txt 

frappe.provide('frappe.core.pages.messages');

frappe.pages.messages.onload = function(wrapper) {
	frappe.ui.make_app_page({
		parent: wrapper,
		title: "Messages"
	});
	
	$('<div><div class="avatar avatar-large">\
		<img id="avatar-image" src="assets/frappe/images/ui/avatar.png"></div>\
		<h3 style="display: inline-block" id="message-title">Everyone</h3>\
	</div><hr>\
	<div id="post-message">\
	<textarea class="form-control" rows=3 style="margin-bottom: 15px;"></textarea>\
	<div><button class="btn btn-default">Post</button></div><hr>\
	</div>\
	<div class="all-messages"></div><br>').appendTo($(wrapper).find('.layout-main-section'));

	wrapper.appframe.add_module_icon("Messages");
	
	frappe.core.pages.messages = new frappe.core.pages.messages(wrapper);
}

$(frappe.pages.messages).bind('show', function() {
	// remove alerts
	$('#alert-container .alert').remove();
	
	frappe.core.pages.messages.show();
	setTimeout("frappe.core.pages.messages.refresh()", 5000);
})

frappe.core.pages.messages = Class.extend({
	init: function(wrapper) {
		this.wrapper = wrapper;
		this.show_active_users();
		this.make_post_message();
		this.make_list();
		//this.update_messages('reset'); //Resets notification icons		
	},
	make_post_message: function() {
		var me = this;
		
		$('#post-message .btn').click(function() {
			var txt = $('#post-message textarea').val();
			if(txt) {
				return frappe.call({
					module: 'frappe.core',
					page:'messages',
					method:'post',
					args: {
						txt: txt,
						contact: me.contact
					},
					callback:function(r,rt) {
						$('#post-message textarea').val('')
						me.list.run();
					},
					btn: this
				});
			}			
		});
	},
	show: function() {
		var contact = this.get_contact() || this.contact || user;

		$('#message-title').html(contact===user ? "Everyone" :
			frappe.user_info(contact).fullname)

		$('#avatar-image').attr("src", frappe.utils.get_file_link(frappe.user_info(contact).image));

		$("#show-everyone").toggle(contact!==user);
		
		$("#post-message button").text(contact==user ? __("Post Publicly") : __("Post to user"))
		
		this.contact = contact;
		this.list.opts.args.contact = contact;
		this.list.run();
		
	},
	// check for updates every 5 seconds if page is active
	refresh: function() {
		setTimeout("frappe.core.pages.messages.refresh()", 5000);
		if(frappe.container.page.label != 'Messages') 
			return;
		if(!frappe.session_alive) 
			return;
		this.show();
	},
	get_contact: function() {
		var route = location.hash;
		if(route.indexOf('/')!=-1) {
			var name = decodeURIComponent(route.split('/')[1]);
			if(name.indexOf('__at__')!=-1) {
				name = name.replace('__at__', '@');
			}
			return name;
		}
	},
	make_list: function() {
		this.list = new frappe.ui.Listing({
			parent: $(this.wrapper).find('.all-messages'),
			method: 'frappe.core.page.messages.messages.get_list',
			args: {
				contact: null
			},
			hide_refresh: true,
			no_loading: true,
			render_row: function(wrapper, data) {
				$(wrapper).removeClass('list-row');
				
				data.creation = dateutil.comment_when(data.creation);
				data.comment_by_fullname = frappe.user_info(data.owner).fullname;
				data.image = frappe.utils.get_file_link(frappe.user_info(data.owner).image);
				data.mark_html = "";

				data.reply_html = '';
				if(data.owner==user) {
					data.cls = 'message-self';
					data.comment_by_fullname = 'You';	
				} else {
					data.cls = 'message-other';
				}

				// delete
				data.delete_html = "";
				if(data.owner==user || data.comment.indexOf("assigned to")!=-1) {
					data.delete_html = repl('<a class="close" \
						onclick="frappe.core.pages.messages.delete(this)"\
						data-name="%(name)s">&times;</a>', data);
				}
				
				if(data.owner==data.comment_docname && data.parenttype!="Assignment") {
					data.mark_html = "<div class='message-mark' title='Public'\
						style='background-color: green'></div>"
				}

				wrapper.innerHTML = repl('<div class="message %(cls)s">%(mark_html)s\
						<span class="avatar avatar-small"><img src="%(image)s"></span><b>%(comment)s</b>\
						%(delete_html)s\
						<div class="help">by %(comment_by_fullname)s, %(creation)s</div>\
					</div>\
					<div style="clear: both;"></div>', data);
			}
		});
	},
	delete: function(ele) {
		$(ele).parent().css('opacity', 0.6);
		return frappe.call({
			method: 'frappe.core.page.messages.messages.delete',
			args: {name : $(ele).attr('data-name')},
			callback: function() {
				$(ele).parent().toggle(false);
			}
		});
	},
	show_active_users: function() {
		var me = this;
		return frappe.call({
			module:'frappe.core',
			page:'messages',
			method:'get_active_users',
			callback: function(r,rt) {
				var $body = $(me.wrapper).find('.layout-side-section');
				$('<h4>Users</h4><hr>\
					<div id="show-everyone">\
						<a href="#messages/'+user+'" class="btn btn-default">\
							Messages from everyone</a><hr></div>\
				').appendTo($body);

				$("#show-everyone").toggle(me.contact!==user);
				
				r.message.sort(function(a, b) { return b.has_session - a.has_session; });
				for(var i in r.message) {
					var p = r.message[i];
					if(p.name != user) {
						p.fullname = frappe.user_info(p.name).fullname;
						p.image = frappe.utils.get_file_link(frappe.user_info(p.name).image);
						p.name = p.name.replace('@', '__at__');
						p.status_color = p.has_session ? "green" : "#ddd";
						p.status = p.has_session ? "Online" : "Offline";
						$(repl('<p>\
							<span class="avatar avatar-small" \
								title="%(status)s"><img src="%(image)s" /></span>\
							<a href="#!messages/%(name)s">%(fullname)s</a>\
							</p>', p))
							.appendTo($body);						
					}
				}
			}
		});
	}
});


